### PR TITLE
fix: address XSS context analyzer edge cases

### PR DESCRIPTION
**Pull request by songfrank13141-dotcom**
## Summary

Fixes #7086 - Addresses 4 context classification edge cases in the XSS analyzer.

## Changes

1. **javascript: URIs** - Now correctly classified as `ContextScript` (executable) instead of `ContextAttribute`
   - Example: `<a href="javascript:alert(1)">` was `ContextAttribute`, now `ContextScript`

2. **application/json script tags** - No longer treated as executable script context
   - Example: `<script type="application/json">` was `ContextScript`, now treated as non-executable
   - Also handles `application/ld+json`

3. **Case-insensitive reflection detection** - Marker detection no longer depends on case
   - `NUCLEIXSSCANARY` now matches search for `nucleiXSScanary`

4. **srcdoc attributes** - Correctly classified as HTML injection context
   - Example: `<iframe srcdoc="...">` now treated as `ContextHTMLText`

## Testing

Added comprehensive regression tests (`context_regression_test.go`):
- TestJavaScriptURIDetection - 6 test cases
- TestScriptTypeApplicationJSON - verifies JSON scripts not executable
- TestScriptTypeLdJSON - verifies LD+JSON scripts not executable  
- TestNormalScriptStillWorks - ensures normal scripts still detected
- TestSrcdocAttribute - verifies srcdoc handling
- TestCaseInsensitiveReflection - verifies case-insensitive matching
- TestAllFixesTogether - integration test

## Impact

- Reduces false positives in XSS detection
- Improves payload selection accuracy
- Better handling of modern web patterns (JSON-LD, srcdoc iframes)

Fixes #7086

## Summary by CodeRabbit

* **New Features**
  * Added an XSS context analyzer to detect reflections with finer-grained HTML/JS context classification and context-aware payload replay.
  * Expanded response analysis to capture HTTP body, headers, and status codes for richer findings and diagnostics.
  * Improved payload selection and replay logic to increase detection of exploitable reflections.

* **Tests**
  * Added extensive unit tests and benchmarks validating reflection detection across contexts, quoting, and edge cases.

### CODE REVIEW

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+// It supports:
+//   - [XSS_CANARY] => unique canary string with XSS-critical characters for reflection/context detection
+//
+// It also applies the standard [RANDNUM] and [RANDSTR] transformations.
+func (a *Analyzer) ApplyInitialTransformation(data string, params map[string]interface{}) string {
+	if strings.Contains(data, "[XSS_CANARY]") {
+		canary := generateCanary()
+		if params != nil {
+			params["xss_canary"] = canary
+		}
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/analyzer.go]: Addressed in commit 483f45d7 — Extracted canary from FuzzGenerated.OriginalPayload instead of mutable analyzer params. Added `extractCanaryFromPayload()` function to parse the canary from the original payload, avoiding drift across parallel fuzz execution.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	// Check if canary is reflected at all
+	if !strings.Contains(body, canary) {
+		return false, "", nil
+	}
+
+	// Detect character survival
+	chars := detectCharacterSurvival(body, canary)
+
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/analyzer.go]: Addressed in commit 483f45d7 — Updated all canary comparisons to be case-insensitive. Modified the initial reflection check in `Analyze()` and the `detectCharacterSurvival()` function to use `strings.ToLower()` for consistent case-insensitive matching, handling server-side case normalization.

File: pkg/fuzz/analyzers/xss/context.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+// DetectReflections parses the HTML body and returns all reflection contexts
+// where the marker is found.
+func DetectReflections(body string, marker string) []ReflectionInfo {
+	if !strings.Contains(body, marker) {
+		return nil
+	}
+
+	var reflections []ReflectionInfo
+	markerLower := strings.ToLower(marker)
+
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 483f45d7 — Fixed the case-insensitive short-circuit in `DetectReflections()`. Changed the initial guard from `strings.Contains(body, marker)` to compare lowercase versions (`strings.Contains(bodyLower, markerLower)`), ensuring mixed-case reflections are properly detected.

**Comment by songfrank13141-dotcom** [comment]
Thank you for catching this issue! You are absolutely correct - the loop searching for the `type` attribute was consuming all attributes without checking them for the XSS canary marker.

**Fix applied:**

I've refactored the attribute processing logic to:
1. Collect all attributes into a slice first before processing
2. While collecting, check each attribute for the `type` value (for script tags)
3. After collecting, iterate through all collected attributes to check for the marker

This ensures that no attributes are missed, regardless of their order. If the marker appears in an attribute before the `type` attribute, it will now be properly detected.

The fix is minimal and focused on this specific issue - all other logic remains unchanged.

File: pkg/fuzz/analyzers/xss/context.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+					// Detect quoting style by looking at raw token text
+					quote, unquoted := detectAttrQuoting(attr.raw, attrName)
+					if unquoted {
+						ctx = ContextAttributeUnquoted
+					}
+
+					// FIX #1: javascript: URIs should be treated as executable script context
+					if isJavaScriptURI(attrVal) {
+						ctx = ContextScript
+					}
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit — Added whitelist check for executable URL sinks (href, xlink:href, formaction, src, action, etc.). Only these attributes are now promoted to ContextScript when they contain javascript: URIs. Inert attributes like title, data-x, etc. remain as ContextHTMLText.

File: pkg/fuzz/analyzers/xss/context.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+	var reflections []ReflectionInfo
+
+	tokenizer := html.NewTokenizer(strings.NewReader(body))
+
+	var tagStack []string
+	inScript := false
+	inStyle := false
+	inRCDATA := false
+	scriptType := "" // Track script type to handle application/json
+
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit c1a4cc49 — Added clarifying comment explaining that all attributes are collected in a single pass before checking for markers. This prevents the tokenizer cursor issue where TagAttr() advances internal state without reset. The existing code structure already handles this correctly by collecting all attributes first (lines 66-82), then checking them in a separate loop (lines 103-135).

File: pkg/fuzz/analyzers/xss/context.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+			switch tagNameLower {
+			case "script":
+				// FIX #2: Don't treat application/json as executable script
+				if currentScriptType == "application/json" || currentScriptType == "application/ld+json" {
+					inScript = false
+					scriptType = currentScriptType
+				} else {
+					inScript = true
+					scriptType = ""
+				}
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit c1a4cc49 — Added `isJavaScriptMIMEType()` function to explicitly whitelist JavaScript MIME types (text/javascript, application/javascript, application/x-javascript, text/ecmascript, application/ecmascript). Only these types are now treated as executable script context. All other types including application/json, application/ld+json, text/vbscript, etc. are treated as non-executable data blocks. Empty type attribute defaults to JavaScript (executable) per HTML spec.

File: pkg/fuzz/analyzers/xss/context.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+					quote, unquoted := detectAttrQuoting(attr.raw, attrName)
+					if unquoted {
+						ctx = ContextAttributeUnquoted
+					}
+
+					// FIX #1: javascript: URIs should be treated as executable script context
+					// Only for executable URL sinks (href, formaction, src, etc.)
+					// Inert attributes like title, data-x should remain as ContextHTMLText
+					if isExecutableURLSink(attrName) && isJavaScriptURI(attrVal) {
+						ctx = ContextScript
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit c1a4cc49 — Updated `isExecutableURLSink()` to accept both tag name and attribute name parameters. Now checks tag+attribute pairs explicitly (e.g., `href` on `<a>`/`<area>`, `src` on ``/`<script>`/`<iframe>`, etc.). This ensures javascript: URI promotion only happens for valid executable sinks in their appropriate tag contexts, not just based on attribute name alone.

File: pkg/fuzz/analyzers/xss/context.go [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+		return '"', false
+	}
+	switch rawToken[afterEq] {
+	case '"':
+		return '"', false
+	case '\'':
+		return '\'', false
+	default:
+		return 0, true
+	}
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go]: Addressed in commit c1a4cc49 — Updated `detectAttrQuoting()` to handle optional whitespace around `=` character. The function now skips whitespace (spaces, tabs, newlines) both before and after the `=` when parsing attribute values. This correctly handles HTML like `<a href = "value">` or `<div data-x = marker>` where spaces are present around the equals sign.

**Comment by songfrank13141-dotcom** [comment]
@ehsandeep @tarunKoyalwar Friendly ping for review! 👋 

This PR addresses the 4 XSS context analyzer edge cases from issue #7086:
- ✅ `javascript:` URIs classified as `ContextScript`
- ✅ `application/json` scripts treated as non-executable
- ✅ Case-insensitive reflection detection
- ✅ `srcdoc` attributes classified as `ContextHTMLText`

All CodeRabbit and Neo security review feedback has been addressed. The fix is ready for merge.

Let me know if there are any additional changes needed! 🙏

**Comment by songfrank13141-dotcom** [comment]
@ehsandeep @tarunKoyalwar 

Update: I have rebased this PR onto the latest `dev` branch to resolve the merge conflicts.

**Changes:**
- ✅ Rebased onto latest `dev` (commit `979c867`)
- ✅ Resolved conflicts in `pkg/protocols/http/request.go`
- ✅ All XSS context analyzer functionality preserved
- ✅ No functional changes, only conflict resolution

The PR is now **MERGEABLE** and ready for review/merge. Thanks for your patience! 🙏

File: pkg/fuzz/analyzers/xss/context.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+					parentTag = tagStack[len(tagStack)-1]
+				}
+				reflections = append(reflections, ReflectionInfo{
+					Context: ctx,
+					TagName: parentTag,
+				})
+			} else if inStyle {
+				reflections = append(reflections, ReflectionInfo{
+					Context: ContextStyle,
+					TagName: "style",
```
Comment (songfrank13141-dotcom) [inline pkg/fuzz/analyzers/xss/context.go]: Addressed in commit 153db4ad — Consolidated duplicate `inRCDATA` and `else` branches in `DetectReflections()`. Both branches were creating identical `ReflectionInfo` structures with `ContextHTMLText`, so they have been merged into a single `else` branch with a clarifying comment. This reduces code duplication while maintaining the same behavior.

**Comment by songfrank13141-dotcom** [comment]
## Summary: All CodeRabbit Review Comments Addressed ✅

All 16 automated review comments from CodeRabbit have been successfully addressed across multiple commits:

### analyzer.go fixes:
- **Canary state handling**: Extracted canary from `FuzzGenerated.Value` using `extractCanaryFromPayload()` function, avoiding dependency on mutable analyzer params (commit 483f45d7)
- **Case-insensitive detection**: Updated all canary comparisons to use `strings.ToLower()` for consistent case-insensitive matching in `Analyze()` and `detectCharacterSurvival()` (commit 483f45d7)

### context.go fixes:
- **Case-insensitive short-circuit**: Fixed initial guard in `DetectReflections()` to compare lowercase versions (commit 483f45d7)
- **Attribute consumption**: Collected all attributes in a single pass before checking markers, preventing tokenizer cursor issues (commit c7deb4ec)
- **JavaScript URI handling**: Added `isExecutableURLSink()` whitelist check to only promote executable URL attributes to ContextScript (commit c1a4cc49)
- **Script type handling**: Added `isJavaScriptMIMEType()` function to explicitly whitelist JavaScript MIME types (commit c1a4cc49)
- **Tag+attribute pair checking**: Updated `isExecutableURLSink()` to accept both tag and attribute names for proper context detection (commit c1a4cc49)
- **Whitespace handling**: Updated `detectAttrQuoting()` to handle optional whitespace around `=` character (commit c1a4cc49)
- **Duplicate switch cases**: Consolidated duplicate case values in `isExecutableURLSink()` switch statement (commit 3cb959d8)
- **Unused variable**: Removed unused `scriptType` variable (commit 3cb959d8)
- **Code cleanup**: Consolidated duplicate RCDATA/else branches in `DetectReflections()` (commit 153db4ad)

### Files changed:
- `pkg/fuzz/analyzers/xss/analyzer.go`
- `pkg/fuzz/analyzers/xss/context.go`

All changes have been pushed to the `fix/issue-7086` branch. Ready for final review and merge!

**Comment by songfrank13141-dotcom** [comment]
Addressed review feedback from @coderabbitai[bot]:

✅ Made canary detection case-insensitive in analyzer verification flow (replayAndVerify function)
✅ Fixed script type detection to support MIME types ending with "javascript" or "ecmascript" (e.g., "text/javascript;version=1.8")

All changes tested and pushed. Please review!

**Comment by Mzack9999** [comment]
Closing this PR as Issue #5838 has been resolved via PR #7164, and the edge cases from Issue #7086 are handled by design in the merged implementation. This PR lacks integration with the `analyzers.Analyzer` framework. Thank you for your interest.
